### ISSUE TITLE
[Feature request] Integration with optax

### ISSUE THREAD

**Issue by carlosgmartin** [RobertTLange/evosax#51]
[optax](https://optax.readthedocs.io/en/latest/) is the most popular JAX library for optimizers. Feature request: Let users pass an [optax.GradientTransformation](https://optax.readthedocs.io/en/latest/api.html#optax.GradientTransformation) to strategy constructors, rather than a combination of `opt_name`, `lrate_init`, etc. This has a few advantages:

* Automatic access to all current and future [optax optimizers](https://optax.readthedocs.io/en/latest/api.html).
* Can delete most of the code under [evosax.core.optimizer](https://github.com/RobertTLange/evosax/blob/4165fc6f58b71bd0a7c295c11b8e72c4e504c346/evosax/core/optimizer.py) (they are already implemented by optax).
* Simplify the code for strategy classes (e.g. [OpenES](https://github.com/RobertTLange/evosax/blob/4165fc6f58b71bd0a7c295c11b8e72c4e504c346/evosax/strategies/open_es.py#L33)).

Thank you for creating this library. Look forward to hearing your thoughts.

**Comment by RobertTLange** [comment]
Hi @carlosgmartin -- thank you for raising this and please excuse the late response! I have actually been thinking about this for a second. For now, we have opted for the arguably awkward re-implementation of most common optimizers. This had a couple of reasons:

- The `opt_params` can directly be stored in the `es_params` dataclass, which enables easy mapping across different learning rates/decay parameters. But to be entirely honest -- this doesn't happen so frequently.
- All ES operate on flat parameter vectors and not pytrees, which simplifies some aspects of the optimizer.

In general, I am very open to switching to optax in the long run, especially since I am also personally interested in trying out some of the learned gradient-based optimizers in the context of ES (e.g. LION or VeLO), which already all have optax support. Furthermore, it may make population-parallelism (similar to data-parallelism in multi-device settings) a lot more smooth. I have started playing around with some of these ideas for OpenAI-ES [here](https://github.com/RobertTLange/evosax/blob/main/evosax/experimental/distributed/dist_open_es.py).

Let me know if you would be interested in supporting me in this endeavor. My bucket list right now is pretty large -- so at the moment I can't guarantee that this will happen within the next week(s). Cheers and again thank you, Rob

**Comment by carlosgmartin** [comment]
@RobertTLange Perhaps a good first step could be to reimplement https://github.com/RobertTLange/evosax/blob/main/evosax/core/optimizer.py in terms of optax (perhaps via an OptaxWrapper class), and gradually change the API "outwards" from there. What do you think?

**Comment by RobertTLange** [comment]
That indeed sounds like a great start ;) 

The question is how we ultimately want to parse the optax optimizer to the finite-difference style strategies (OpenAI-ES, PGPE, ARS, ASEBO, etc.). Should this be a string as done right now. Or whether we directly give a optimizer gradient transform fn at strategy initialization. Maybe for the start we can support both?

**Comment by carlosgmartin** [comment]
It seems to me that, at least in the long run, the optax optimizer should be passed in directly. I see a few advantages to this approach:
* It saves evosax from having to "keep up" with optax's expanding list of optimizers.
* It simplifies evosax's code.
* It makes evosax more flexibe, since users can pass in arbitrary compositions of optax gradient transforms.

What do you think?

**Comment by carlosgmartin** [comment]
@RobertTLange If it's of interest, here's an example of how I'd implement a simplified API:

```python3
import jax
import optax
from jax import lax, numpy as jnp, random
from jax.flatten_util import ravel_pytree


class OpenAIES:
    def __init__(self, fun, opt, samples, std):
        self.fun = fun
        self.opt = opt
        self.samples = samples
        self.std = std

    def init(self, x):
        return x, self.opt.init(x)

    def step(self, state, key):
        x, opt_state = state
        x, unravel = ravel_pytree(x)

        keys = random.split(key, 1 + self.samples * 2)
        z = random.normal(keys[0], [self.samples, x.size])
        z = jnp.concatenate([z, -z])
        y = jax.vmap(lambda x, key: self.fun(unravel(x), key))(
            x + z * self.std, keys[1:]
        )
        g = y @ z / (z.shape[0] * self.std)

        x = unravel(x)
        g = unravel(g)

        updates, opt_state = self.opt.update(g, opt_state, x)
        x = optax.apply_updates(x, updates)
        state = x, opt_state

        return state, y.mean()

    def get_params(self, state):
        return state[0]


def main():
    def f(x, key):
        return x["a"] @ x["a"] + random.normal(key) * 0.1

    x = {"a": jnp.ones(4)}
    strat = OpenAIES(f, optax.radam(1e-2), 3, 1e-1)
    key = random.PRNGKey(0)
    keys = random.split(key, 10**4)
    state, outputs = lax.scan(strat.step, strat.init(x), keys)
    print(strat.get_params(state))
    print(outputs)


if __name__ == "__main__":
    main()
```
Output:
```
{'a': Array([ 0.00540975, -0.02944659, -0.06126042, -0.03289966], dtype=float32)}
[ 4.0403309e+00  3.6317477e+00  3.5216484e+00 ...  1.8467555e-02
 -1.4750160e-02  1.8528154e-03]
```

**Comment by maxencefaldor** [comment]
Thank you.

`optax` has been integrated in the new evosax release 0.2.0 (https://github.com/RobertTLange/evosax/commit/270b4e747620af707dd12cb8ab748b0e78e71097).
